### PR TITLE
MESH-1473/Create a list of transactions that require Asset/Portfolio permissions

### PR DESCRIPTION
**Pull request by CJP10**
### new features

- creates a list of transactions that require Asset/Portfolio permissions

### CODE REVIEW

File: utils/permissions/src/doc_parser/enums.rs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+pub struct Variant {
+    pub name: String,
+    pub doc_block: Html,
+}
+
+pub struct EnumDocParser {
+    html: Html,
+}
+
+impl EnumDocParser {
```
Comment (Centril) [inline utils/permissions/src/doc_parser/enums.rs:24]: Please add some commentary at various "paragraphs" in the code below... :)

File: utils/permissions/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        .map(|(pallet, doc)| (pallet, parse_permissions(doc)))
+        .for_each(|(pallet, permissions)| {
+            for (extrinsic, permission) in permissions {
+                let extrinsic_mapping = permission_mappings
+                    .entry(pallet.clone())
+                    .or_insert(BTreeMap::new());
+                extrinsic_mapping
+                    .entry(extrinsic.clone())
+                    .or_insert(Vec::new())
+                    .push(permission);
```
Comment (Centril) [inline utils/permissions/src/main.rs:59]:
```suggestion
                permission_mappings
                    .entry(pallet.clone())
                    .or_default()
                    .entry(extrinsic.clone())
                    .or_default()
                    .push(permission);
```

File: utils/permissions/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    println!("{:#?}", permission_mappings);
+}
+
+fn parse_permissions(doc: EnumDoc) -> Vec<(String, String)> {
+    let mut permissions = Vec::new();
+    for variant in doc.variants {
+        if let Some(doc_block_div) = variant
+            .doc_block
+            .select(&Selector::parse("div.docblock").unwrap())
+            .next()
```
Comment (Centril) [inline utils/permissions/src/main.rs:72]: Let's fuse these using `.filter_map`

File: utils/permissions/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        {
+            let children: Vec<ElementRef> = doc_block_div
```
Comment (Centril) [inline utils/permissions/src/main.rs:74]:
```suggestion
            let children: Vec<_> = doc_block_div
```

File: utils/permissions/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                .select(&Selector::parse("*").unwrap())
+                .collect();
+
+            let h1_selector = Selector::parse("h1").unwrap();
+            let ul_selector = Selector::parse("ul").unwrap();
+            let li_selector = Selector::parse("li").unwrap();
```
Comment (Centril) [inline utils/permissions/src/main.rs:80]: LLVM is probably clever enough to apply the loop-invariant-code-motion on this to float it out of the loop, but it's good to develop a habit of not relying on it in general, although performance is not important for this application specifically.

File: utils/permissions/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                        .value()
+                        .id()
+                        .map(|s| s == "permissions")
+                        .unwrap_or(false)
+                {
+                    if let Some(ul) = children[i + 1..]
+                        .iter()
+                        .take_while(|element| !h1_selector.matches(element))
+                        .filter(|element| ul_selector.matches(element))
+                        .next()
```
Comment (Centril) [inline utils/permissions/src/main.rs:94]: Prefer `.filter` + `.filter_map`

File: utils/permissions/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,108 @@
+use crate::doc_parser::enums::{EnumDoc, EnumDocParser};
```
Comment (Centril) [inline utils/permissions/src/main.rs:1]: Let's add more "paragraph"-commentary here as well so that the code's broad strokes are more understandable.

**Review by Centril — CHANGES_REQUESTED** [review]

**Comment by CJP10** [comment]
Current output:
```
{
    "pallet_asset": {
        "add_documents": [
            "Asset",
        ],
        "add_extension": [
            "Asset",
        ],
        "archive_extension": [
            "Asset",
        ],
        "create_asset": [
            "Portfolio",
        ],
        "freeze": [
            "Asset",
        ],
        "issue": [
            "Asset",
            "Portfolio",
        ],
        "make_divisible": [
            "Asset",
        ],
        "redeem": [
            "Asset",
            "Portfolio",
        ],
        "register_ticker": [
            "Asset",
        ],
        "remove_documents": [
            "Asset",
        ],
        "remove_primary_issuance_agent": [
            "Asset",
        ],
        "remove_smart_extension": [
            "Asset",
        ],
        "rename_asset": [
            "Asset",
        ],
        "set_funding_round": [
            "Asset",
        ],
        "unarchive_extension": [
            "Asset",
        ],
        "unfreeze": [
            "Asset",
        ],
        "update_identifiers": [
            "Asset",
        ],
    },
    "pallet_compliance_manager": {
        "add_compliance_requirement": [
            "Asset",
        ],
        "add_default_trusted_claim_issuer": [
            "Asset",
        ],
        "change_compliance_requirement": [
            "Asset",
        ],
        "pause_asset_compliance": [
            "Asset",
        ],
        "remove_compliance_requirement": [
            "Asset",
        ],
        "remove_default_trusted_claim_issuer": [
            "Asset",
        ],
        "replace_asset_compliance": [
            "Asset",
        ],
        "reset_asset_compliance": [
            "Asset",
        ],
        "resume_asset_compliance": [
            "Asset",
        ],
    },
    "pallet_corporate_actions": {
        "change_record_date": [
            "Asset",
        ],
        "initiate_corporate_action": [
            "Asset",
        ],
        "link_ca_doc": [
            "Asset",
        ],
        "remove_ca": [
            "Asset",
        ],
        "reset_caa": [
            "Asset",
        ],
        "set_default_targets": [
            "Asset",
        ],
        "set_default_withholding_tax": [
            "Asset",
        ],
        "set_did_withholding_tax": [
            "Asset",
        ],
    },
    "pallet_portfolio": {
        "delete_portfolio": [
            "Portfolio",
        ],
        "move_portfolio_funds": [
            "Portfolio",
        ],
        "rename_portfolio": [
            "Portfolio",
        ],
    },
    "pallet_settlement": {
        "add_and_affirm_instruction": [
            "Portfolio",
        ],
        "affirm_instruction": [
            "Portfolio",
        ],
        "affirm_with_receipts": [
            "Portfolio",
        ],
        "allow_venues": [
            "Asset",
        ],
        "claim_receipt": [
            "Portfolio",
        ],
        "disallow_venues": [
            "Asset",
        ],
        "reject_instruction": [
            "Portfolio",
        ],
        "set_venue_filtering": [
            "Asset",
        ],
        "unclaim_receipt": [
            "Portfolio",
        ],
        "withdraw_affirmation": [
            "Portfolio",
        ],
    },
    "pallet_statistics": {
        "add_exempted_entities": [
            "Asset",
        ],
        "add_transfer_manager": [
            "Asset",
        ],
        "remove_exempted_entities": [
            "Asset",
        ],
        "remove_transfer_manager": [
            "Asset",
        ],
    },
    "pallet_sto": {
        "create_fundraiser": [
            "Asset",
            "Portfolio",
        ],
        "freeze_fundraiser": [
            "Asset",
        ],
        "invest": [
            "Portfolio",
        ],
        "modify_fundraiser_window": [
            "Asset",
        ],
        "stop": [
            "Asset",
        ],
        "unfreeze_fundraiser": [
            "Asset",
        ],
    },
}
```

**Comment by CJP10** [comment]
New output

```
pallet_asset,add_documents,Asset
pallet_asset,add_extension,Asset
pallet_asset,archive_extension,Asset
pallet_asset,create_asset,Portfolio
pallet_asset,freeze,Asset
pallet_asset,issue,Asset,Portfolio
pallet_asset,make_divisible,Asset
pallet_asset,redeem,Asset,Portfolio
pallet_asset,register_ticker,Asset
pallet_asset,remove_documents,Asset
pallet_asset,remove_primary_issuance_agent,Asset
pallet_asset,remove_smart_extension,Asset
pallet_asset,rename_asset,Asset
pallet_asset,set_funding_round,Asset
pallet_asset,unarchive_extension,Asset
pallet_asset,unfreeze,Asset
pallet_asset,update_identifiers,Asset
pallet_compliance_manager,add_compliance_requirement,Asset
pallet_compliance_manager,add_default_trusted_claim_issuer,Asset
pallet_compliance_manager,change_compliance_requirement,Asset
pallet_compliance_manager,pause_asset_compliance,Asset
pallet_compliance_manager,remove_compliance_requirement,Asset
pallet_compliance_manager,remove_default_trusted_claim_issuer,Asset
pallet_compliance_manager,replace_asset_compliance,Asset
pallet_compliance_manager,reset_asset_compliance,Asset
pallet_compliance_manager,resume_asset_compliance,Asset
pallet_corporate_actions,change_record_date,Asset
pallet_corporate_actions,initiate_corporate_action,Asset
pallet_corporate_actions,link_ca_doc,Asset
pallet_corporate_actions,remove_ca,Asset
pallet_corporate_actions,reset_caa,Asset
pallet_corporate_actions,set_default_targets,Asset
pallet_corporate_actions,set_default_withholding_tax,Asset
pallet_corporate_actions,set_did_withholding_tax,Asset
pallet_portfolio,delete_portfolio,Portfolio
pallet_portfolio,move_portfolio_funds,Portfolio
pallet_portfolio,rename_portfolio,Portfolio
pallet_settlement,add_and_affirm_instruction,Portfolio
pallet_settlement,affirm_instruction,Portfolio
pallet_settlement,affirm_with_receipts,Portfolio
pallet_settlement,allow_venues,Asset
pallet_settlement,claim_receipt,Portfolio
pallet_settlement,disallow_venues,Asset
pallet_settlement,reject_instruction,Portfolio
pallet_settlement,set_venue_filtering,Asset
pallet_settlement,unclaim_receipt,Portfolio
pallet_settlement,withdraw_affirmation,Portfolio
pallet_statistics,add_exempted_entities,Asset
pallet_statistics,add_transfer_manager,Asset
pallet_statistics,remove_exempted_entities,Asset
pallet_statistics,remove_transfer_manager,Asset
pallet_sto,create_fundraiser,Asset,Portfolio
pallet_sto,freeze_fundraiser,Asset
pallet_sto,invest,Portfolio
pallet_sto,modify_fundraiser_window,Asset
pallet_sto,stop,Asset
pallet_sto,unfreeze_fundraiser,Asset
```

**Review by maxsam4 — APPROVED** [review]

**Review by vkomenda — APPROVED** [review]

File: utils/permissions/src/doc_parser/enums.rs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        let all_divs: Vec<ElementRef> = self.html.select(&all_div_selector).collect();
+
+        all_divs
+            .iter()
+            .enumerate()
+            .filter_map(|(i, _)| self.parse_variant(&all_divs, i))
+            .collect()
+    }
+
+    fn parse_variant(&self, all_divs: &Vec<ElementRef>, index: usize) -> Option<Variant> {
```
Comment (fmiguelgarcia) [inline utils/permissions/src/doc_parser/enums.rs:43]:
```suggestion
    fn parse_variant(&self, all_divs: &[ElementRef], index: usize) -> Option<Variant> {
```

**Review by fmiguelgarcia — APPROVED** [review]

**Review by satyamakgec — APPROVED** [review]

**Pull request merged**